Query Property Wrappers, HealthChart, and Other Refactoring

lukaskollmer opened the pull request:
# query property wrappers, HealthChart, other refactoring

## :recycle: Current situation & Problem
The HealthKit module currently only provides access to HealthKit data via long-running anchor queries that deliver information about new/deleted objects to the app's Standard.
It does not provide any facilities for querying for past samples, or accessing health data from SwiftUI. This PR attempts to address these issues.

Furthermore, the HealthKit module is lacking an API allowing spezi users to integrate custom HealthKit permission requests into the module's permission handling (i.e., you currently can only request HealthKit access for some specific sample type by actively defining a long-running observer for that sample type).

Furthermore, this PR attempts to implement a `HealthChart` view, which can display various types of queried HealthKit data as a chart.

resolves #8 
requires https://github.com/StanfordSpezi/SpeziFoundation/pull/19
requires https://github.com/StanfordBDHG/XCTestExtensions/pull/28


## :gear: Release Notes 
- Added `HealthKitQuery` property wrapper
- Added `HealthKitStatisticsQuery` property wrapper
- Added `HealthKitCharacteristicQuery` property wrapper
- Added `HealthChart` view
- Extended `HealthKit` configuration API to allow users to specify sample types the system should request read and/or write access to
- Removed `CollectSamples`. The same functionality can be achieved using a `for` loop creating individual `CollectSample` instances.


## :books: Documentation
All added new and changed existing APIs are documented. The DocC structure was reworked and some key aspects of SpeziHealthKit now have dedicated article/extension pages.


## :white_check_mark: Testing
The new and changed APIs are tested using both "normal" unit tests and UI tests. Existing tests were kept and adapted where possible


## :pencil: Code of Conduct & Contributing Guidelines 

By submitting creating this pull request, you agree to follow our [Code of Conduct](https://github.com/StanfordSpezi/.github/blob/main/CODE_OF_CONDUCT.md) and [Contributing Guidelines](https://github.com/StanfordSpezi/.github/blob/main/CONTRIBUTING.md):
- [x] I agree to follow the [Code of Conduct](https://github.com/StanfordSpezi/.github/blob/main/CODE_OF_CONDUCT.md) and [Contributing Guidelines](https://github.com/StanfordSpezi/.github/blob/main/CONTRIBUTING.md).

PSchmiedmayer commented on Package.swift:
```diff
@@ -22,22 +22,28 @@ let package = Package(
+                .product(name: "Spezi", package: "Spezi"),
+                .product(name: "SpeziFoundation", package: "SpeziFoundation")
             ],
+            swiftSettings: [.enableUpcomingFeature("ExistentialAny")],
```
🚀

PSchmiedmayer commented on Sources/SpeziHealthKit/CollectSample/CollectSample.swift:
```diff
@@ -10,11 +10,14 @@ import HealthKit
 ///
-/// This structure define what and how the ``HealthKit`` samples are collected. By default, all samples of the provided `HKSampleType` will be collected. The collection starts on calling ``HealthKit/triggerDataSourceCollection()`` if you configure the `deliverySetting` as    ``HealthKitDeliverySetting/manual(safeAnchor:)`` or automatic once the application is launched when you configure anything  ...
+/// This structure define what and how the ``HealthKit`` samples are collected. By default, all samples of the provided ``HealthKitSampleType`` will be collected.
+/// The collection starts on calling ``HealthKit/triggerDataSourceCollection()`` if you configure the `deliverySetting` as    ``HealthKitDeliverySetting/manual(saveAnchor:)`` or automatic once the application is launched when you configure anything else than manual, i.e.  ``HealthKitDeliverySetting/anchorQuery(_:saveAnchor:)`` or ``HealthKitDeliverySetting/background(_:saveAnchor:)``.
```
We might want to clarify here that (automatic) collection only starts once the permissions have been requested authorization. Maybe we should add a DocC hint/tip box here to point to the function to request authorization from HealthKit?

PSchmiedmayer commented on Sources/SpeziHealthKit/CollectSample/CollectSample.swift:
```diff
@@ -38,39 +41,55 @@ import Spezi
-    
-    public var sampleTypes: Set<HKSampleType> {
-        collectSamples.sampleTypes
+    public var dataAccessRequirements: HealthKitDataAccessRequirements {
```
Nice addition!

PSchmiedmayer commented on Sources/SpeziHealthKit/Sample Types/SampleType.swift:
```diff
@@ -0,0 +1,199 @@
+
+
+
+// TODO is this actually needed?
```
Note to address this TODO before merging the PR.

PSchmiedmayer commented on Sources/SpeziHealthKit/Sample Types/SampleType.swift:
Nice mechanism; very cool!

PSchmiedmayer commented on Sources/SpeziHealthKit/Sample Types/SampleType+QuantityTypes.swift:
```diff
@@ -0,0 +1,66 @@
+        .heartRate,
+        displayTitle: "Heart Rate",
+        displayUnit: .count() / .minute()
+//        expectedValuesRange: 60...150
```
I like the idea of directly injecting the expected ranges here.

lukaskollmer replied:
ideally, i'd have the chart simply look at the min/max of the values it's trying to plot and auto-adjust its y axis based on that.

the issue is that this is actually only sometimes what you want: if you e.g. have heart rate samples and some of them exceed the "normal" range of heart rates, you want the chart to adjust to that. but if you e.g. only have heart rates in the range `90...94`, you don't want to get a chart with a y axis that only covers a range of like 10 BPM.

i'll keep and properly implement the `expectedValuesRange` approach

PSchmiedmayer commented on Sources/SpeziHealthKit/Sample Types/SampleType.swift:
```diff
@@ -0,0 +1,199 @@
+
+// MARK: Factory methods for commonly-used sample types
+
+extension HealthKitSampleType {
```
We might want to make these public if we don't expect to create individual static values for all quantity types (and even then it might be good to have an escape hatch if needed when new sample types are added).

PSchmiedmayer commented on Sources/SpeziHealthKit/Sample Types/SampleType+QuantityTypes.swift:
```diff
@@ -0,0 +1,66 @@
+import HealthKit
+
+
+extension HealthKitSampleType where Sample == HKQuantitySample {
```
How extensible to we plan to be here; do we add all currently known sample types here?

lukaskollmer replied:
(regarding this and the question whether the factory methods should be public)

i've been thinking about this as well, but i'm not sure what the best course of action would be.

ideally, i'd imagine the type as being defined in a way that you just get a bunch of extensions for everything you might need and work with that (kinda like how SwiftUI works, e.g. when looking at `ButtonStyle` or `FormStyle`).
i was kinda worried that, if we had e.g. `SampleType.heartRate` and `SampleType.quantity(identifier:...)`, users would maybe not use the pre-defined one and define everything on their own.

i probably would've ended up defining static properties for all currently known sample types (or, if there's issues with some of them, the subset of sample types we currently can support).

that being said: i can see the appeal of allowing users to define custom sample types.
firstly, it would make it easier to work with new types added by Apple.
secondly, it would allow users to override our defaults, e.g. the `expectedValuesRange`, if they want to use something custom.

sample types are intentionally already implemented to use only their internal `HKSampleType` as their identity (for hashing/equality purposes), so it would actually be possible to allow them to be externally defined and still work well regardless of the parameters the user passes in.

PSchmiedmayer commented on Sources/SpeziHealthKit/Sample Types/SampleType.swift:
```diff
@@ -0,0 +1,199 @@
+    
+    /// Creates a new ``HealthKitSampleType``.
+    /// - Note: Don't use this initializer directly. Instead, use one of the static factory methods: `HealthKitSampleType.quantity`, `.correlation`, and `.category`.
+    private init(_ hkSampleType: Sample._SampleType, displayTitle: String, variant: Variant) {
```
Wondering if the the display title should rather be a `LocalizedStringResource` or similar localizable elements to enable this in the future or at least provide the option to use the displayTitles in the UI?

lukaskollmer replied:
yes, definitely.

there's multiple places where the whole `HealthKitSampleType` should get localized:
- anything that is user-displayed text (e.g. the displayTitles of the sample types)
- anything that displays a value in a unit (e.g., need to differentiate between US, metric, and the UK)

i'm currently already doing this a little bit in the `displayUnit` of the `height` sample type (though that is probably still a little wrong, since you can only express feet as a `HKUnit`, but not feet and inches).
once we add a couple more sample types, we'd also need to localize temperatures between C/F/etc.

PSchmiedmayer commented on Sources/SpeziHealthKit/HealthKit Extensions/HKElectrocardiogram+RelatedData.swift:
```diff
@@ -19,60 +19,52 @@ extension HKElectrocardiogram {
-                    continue
+        switch symptomsStatus {
+        case .present:
+            // TODO ideally, the auth reques should be moved out of here!!!
+            // instead, have it run as part of the HealthKit config step?!
```
Agree, maybe correlated identifies could be passed in the configuration options for the ECG HealthKit samples configuration object? Maybe with a default list of all symptoms and a default value that doesn't present them in addition to the ECG data type on authorization request?

lukaskollmer replied:
We already have a mechanism where you can `RequestReadAccess` to eg the blood pressure correlation type (which isn't actually allowed by HealthKit), and it'll automatically transform that into requests for the systolic and diastolic quantity types (which is allowed); we should probably adopt the same approach for this.

lukaskollmer replied:
i didn't end up implementing this; in the current PR bc it'd probably require a non-trivial amount of changes. i'll open an issue for this and fix it in a subsequent PR.

PSchmiedmayer commented on Sources/SpeziHealthKit/HealthKit.swift:
```diff
@@ -68,129 +69,215 @@ import SwiftUI
+            dataReqs.merging(with: component.dataAccessRequirements)
+        }
+        healthKitDataAccessRequirements = _initialConfigHealthKitDataAccessRequirements
+        if !HKHealthStore.isHealthDataAvailable() {
```
Sounds like a reasonable approach.

PSchmiedmayer commented on Sources/SpeziHealthKit/HealthKit.swift:
```diff
@@ -68,129 +69,215 @@ import SwiftUI
-    /// Call this function when you want to start HealthKit data collection.
+    /// - Note: There is no need for an app itself to keep track of whether it already requested health data access; the ``HealthKit-swift.class`` takes care of this.
+    ///
+    /// - Important: You need to call this function at some point during your app's lifecycle, ideally soon after the app was launched.
```
Do I understand this correctly that this always has to be called on every app launch instead of just once for the onboarding of a user and all subsequent launches start to collect the data automatically?

lukaskollmer replied:
Yes.

In a sense, this was already a requirement before, but it was never stated explicitly.
(Imagine e.g. adding a new `CollectSample` to the config, with a sample type you haven't requested read access for yet, or in general needing read access to some new sample type somewhere in the app. You would've needed to call `askForAuthorization()` outside your onboarding to trigger the HealthKit aurhorization screen in that scenario, so putting the call only into the onboarding would only have worked if you never wanted to change anything.)

IMO, this new approach (of always requiring the call) is cleaner overall, since the app doesn't need to keep track of (or somehow determine) whether the user was already asked, which is already handled by the HealthKit module.

lukaskollmer replied:
Though actually now that I think a little bit more about it, it would of course in theory be possible to auto-start data collection for all non-manual, already-authorized data sources. Though this could lead to a situation where an app forgets that it needs to also call `askForAuthorization()` outside its initial onboarding, and then will never receive data for any new/changed config queries.

PSchmiedmayer commented on Sources/SpeziHealthKit/HealthKit.swift:
```diff
@@ -68,129 +69,215 @@ import SwiftUI
+
+
+extension HKHealthStore {
+    // TODO remove?!
```
What would be the reason for this function?

lukaskollmer replied:
It'll get removed, but there are some object types where you need to request per-object access (instead of a simple one-time request covering all objects of the type).
we're currently not using this; i was thinking about maybe adding support but i then concluded that it'd exceed the scope of the PR and probably (right now, at least) also of the HealthKit package as a whole.

i had some issues calling the async version of that function and added a custom overload to work around that.

PSchmiedmayer commented on Sources/SpeziHealthKit/Export.swift:
```diff
@@ -6,4 +6,5 @@
 // SPDX-License-Identifier: MIT
 //
 
+@_documentation(visibility: internal)
```
Thank you!

PSchmiedmayer commented on Sources/SpeziHealthKit/CollectSample/HealthKitSampleDataSource.swift:
```diff
@@ -145,14 +142,12 @@ final class HealthKitSampleDataSource: HealthKitDataSource {
     private func anchoredContinuousObjectQuery() async throws {
-        try await healthStore.requestAuthorization(toShare: [], read: [sampleType])
-        
+        print("ANCHOR", anchor)
```
Print result might be removed or moves into a log statement.

PSchmiedmayer commented on Sources/SpeziHealthKit/CollectSample/HealthKitSampleDataSource.swift:
```diff
@@ -162,31 +157,32 @@ final class HealthKitSampleDataSource: HealthKitDataSource {
                 for addedSample in results.addedSamples {
                     await standard.add(sample: addedSample)
                 }
+                print("NEW ANCHOR", results.newAnchor)
```
Print result might be removed or moves into a log statement.

PSchmiedmayer commented on Sources/SpeziHealthKit/CollectSample/CollectSample.swift:
```diff
@@ -38,39 +41,55 @@ import Spezi
+    public init(
+        _ sampleType: HealthKitSampleType<some Any>,
+        predicate: NSPredicate? = nil,
+        delivery: HealthKitDeliverySetting = .manual() // TODO Question @Paul : why does it default to manual?
```
Because we didn't really have a good read/write setup before. I would argue that anchorQuery might make more sense now that we have a better mechanism to request dedicated read and write access for samples.

lukaskollmer replied:
@PSchmiedmayer now that the auth request handling is centralized, and multiple `CollectSample` configs trying to start automatically at the same time works properly; should we change the default here to `.automatic`?

PSchmiedmayer commented on Sources/SpeziHealthKit/HealthKit.swift:
```diff
@@ -19,46 +20,46 @@ import SwiftUI
 ///         Configuration(standard: ExampleStandard()) {
 ///             if HKHealthStore.isHealthDataAvailable() {
 ///                 HealthKit {
 ///                     CollectSample(
```
We might want to add some examples of the read and write configuration elements here.

PSchmiedmayer commented on Sources/SpeziHealthKit/HealthKit.swift:
```diff
@@ -19,46 +20,46 @@ import SwiftUI
 /// class ExampleAppDelegate: SpeziAppDelegate {
 ///     override var configuration: Configuration {
 ///         Configuration(standard: ExampleStandard()) {
 ///             if HKHealthStore.isHealthDataAvailable() {
```
As far as I can see this could be removed here now that we just provide empty queries if HealthKit is not available.

PSchmiedmayer commented on Sources/SpeziHealthKit/HealthChart/HealthChart.swift:
Very cool mechanism, looking forward to seeing this merged & extended in future iterations. Looks nice, might be worth splitting it up into multiple files as we finalize the implementation.

lukaskollmer replied:
It'll definitely get split up into multiple files, though I'm not actually sure how much leeway i have on this.
There's currently some weird compilation issues around here which seem to also be somehow related to file structure (see eg swiftlang/swift/issues/78734 and swiftlang/swift/issues/78735).
That's also why I sadly had to move the HealthChart files into the SpeziHealthKit target, instead of having a separate SpeziHealthKitUI target (which i would prefer).

PSchmiedmayer replied:
Interesting, that's unfortunate to hear. Thank you for creating the issues in the Swift repo; let's see what they say about this.

Agree that it would be nice to see this moved into a separate target but we might need to wait until the issues are resolved.

PSchmiedmayer commented on Sources/SpeziHealthKit/HealthKitDeliverySetting/HealthKitDeliverySetting.swift:
```diff
@@ -10,14 +10,20 @@
     case background(HealthKitDeliveryStartSetting = .automatic, saveAnchor: Bool = true)
     
+    /// :nodoc:
+    @available(*, unavailable, renamed: "manual(saveAnchor:)")
```
Thank you for fixing this!

PSchmiedmayer commented on Sources/SpeziHealthKit/Queries/HealthKitCharacteristicsQuery.swift:
```diff
@@ -0,0 +1,93 @@
+/// - Note: This property wrapper is not auto-updating; if the characteristic's value is changed while a view using this property wrapper is active,
+///     it will continue displaying the old value until the view gets updated for some other reason.
+@propertyWrapper
+public struct HealthKitCharacteristicQuery<Value>: DynamicProperty {
```
Very nice mechanism; very cool! We should add some nice examples of using this here in the inline docs as well as adjust the README & maybe add some DocC articles with examples how to use this in combination with a minimal possible view.

PSchmiedmayer commented on Sources/SpeziHealthKit/Queries/HealthKitStatisticsQuery.swift:
Very nice, some of the TODOs can subsequently be resolved but the overall structure looks nice, a lot of convenient shorthand elements!

PSchmiedmayer commented on Sources/SpeziHealthKit/Queries/HealthKitQuery.swift:
Very cool mechanism; looking forward to seeing this merged 🚀

PSchmiedmayer reviewed (approved):
Great job with the PR and additions here @lukaskollmer!

I added some comments throughout the PR. Thank you for resolving all the remaining TODOs in this PR or we start to transform some of them into GitHub issues to be resolved later on in subsequent PRs.

We should also adjust the README (similar to other Spezi READMEs) to reflect the functionality & HealthChart examples. We should also copy the same content to the DocC setup & ensure that it also is having all the updated references and examples in place.

This will be a huge step forward for SpeziHealthKit and will bring it closer to a 1.0, thank you! 🚀

lukaskollmer commented on Sources/SpeziHealthKit/Queries/HealthKitQuery.swift:
```diff
@@ -0,0 +1,515 @@
+
+
+/// The time range for which data should be fetched from the health store.
+public enum HealthKitQueryTimeRange: Hashable, Sendable {
```
@PSchmiedmayer i ended up adding probably slightly too many cases here. what's your take as to which time ranges we should want to offer?

PSchmiedmayer commented on Sources/SpeziHealthKit/Sample Types/SampleType+QuantityTypes.swift:
```diff
@@ -9,58 +9,54 @@
 
-extension HealthKitSampleType where Sample == HKQuantitySample {
+/// ## Quantity Sample Types
+extension SampleType {
```
I like the simplification into SampleType 👍

PSchmiedmayer commented on Sources/SpeziHealthKit/Health Data Collection/HealthDataCollectorDeliverySetting.swift:
```diff
@@ -0,0 +1,76 @@
+    case manual(saveAnchor: Bool = true)
+    /// The HealthKit data is collected based on the ``Start`` and constantly listens to updates while the application is running.
+    /// If `saveAnchor` is enabled the `HKQueryAnchor` is persisted across multiple application launches using the user defaults.
+    case anchorQuery(Start = .automatic, saveAnchor: Bool = true)
```
As the manual setting now also includes the option to save the anchor we might want to rename this something like `continuousQuery`, `continuous`, or something similar? As we don't name `background` `backgroundQuery`, it might make more sense to have `manual`, `continuous`, and `background`?

PSchmiedmayer replied:
Good points, we should ensure that the terminology is aligned across all parts of the package.

I don't think I have better names than `manual`, `continuous`, and `background` at this point but we can continue to think about them as we get closer to a 1.0 release. We might want to merge the PR as it is sooner than later to avoid larger and larger diffs.

lukaskollmer replied:
agreed. my main focus at the moment is to clean up the implementation, restructure the files a bit, write tests and then get it merged.
IMO, the public API and feature set is currently in a good state and there aren't any changes i want to make in that regard (though i might end up removing some functionality from the HealthChart, so that the first version is as stable and usable as possible; it can always get added back later)

lukaskollmer replied:
I haven't renamed these yet, but i think calling it `continuous`, `background`, and `manual` is good. i'll change it to that before the PR is merged, and add deprecated aliases so that old code can migrate 👍

PSchmiedmayer reviewed (approved):
Very cool job with the PR; happy to see this merged in a first version and we can try to refine and improve things as we integrate it into a few first application using some beta tags (E.g. 1.0.0-beta.1) to get started.

Great job; amazing PR 🚀

lukaskollmer commented:
@PSchmiedmayer a little update so that you don't have to look over the entire diff:
- i wrote documentation, DocC articles, and extension pages for all new things, and restructured the existing documentation a bit
  - (the `CollectSample` page now has a neat little table which explains how the different delivery options affect how/when the data collection starts)
- i incorporated your TimeRange struct suggestion
- i had to significantly reduce the HealthChart's feature set (there were weird compiler crashes, and i'd rather attempt to fix that in a separate PR)
  - i removed the ability to have a long-press highlight, or enable scrolling (we never supported both at the same time, bc of limitations in Swift Charts, but for now they're both gone)
  - i removed the `makeInteractive` modifier which added a segmented control where you could choose between day/month/etc for the chart's time range
- i'm currently working on the tests; the plan is to have both "normal" tests and UITests via the automated app
  - i'm not yet sure how well i can test HealthKit, since that's not really easy to mock; i'll see what i can do on that

PSchmiedmayer commented on CONTRIBUTORS.md:
```diff
@@ -13,3 +13,4 @@ SpeziHealthKit contributors
 
 * [Paul Schmiedmayer](https://github.com/PSchmiedmayer)
 * [Andreas Bauer](https://github.com/Supereg)
+* [Lukas Kollmer](https://github.com/lukaskollmer)
```
🚀

PSchmiedmayer reviewed (approved):
Amazing; thank you @lukaskollmer! I looked over the diff and I am happy to see the changes merged; no larger feedback at this point. Feel free to merge this once all the CI elements are passing.

PSchmiedmayer reviewed (approved):
Amazing job, thank you! 🚀
